build.py

```
@@ -0,0 +1,418 @@
+            ]
+        )
+
+    def assemble_ssh_command(self):
+        return " ".join(
+            [
+                f"ssh",
+                f"-i",
+                f"{self.pvt_key}",
+                f"{SSH_USER}@{SSH_HOST}",
+                f"chmod",
+                f"ugo+r",
+                f"{STATIC_DIR}{os.path.basename(self.image_path)}",
+            ]
+        )
+
+    def assemble_timestamp(self):
+        today = datetime.date.today()
+        seconds_since_midnight = time.time() - time.mktime(today.timetuple())
+        return today.strftime("%Y%m%d") + "~" + str(int(seconds_since_midnight))
+
+    def build(self):
+        self.clean_image_dir()
+        run_subprocess_with_spinner(
+            "BUILD",
+            subprocess.Popen(
+                self.assemble_packer_build_command(),
+                env=self.assemble_packer_envs(),
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+                close_fds=True,
+                shell=True,
+            ),
+        )
+
+    def convert(self):
+        run_subprocess_with_spinner(
+            name="CONVERT",
+            proc=subprocess.Popen(
+                self.assemble_convert_command(),
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+                close_fds=True,
+                shell=True,
+            ),
+        )
+
+    def clean_image_dir(self):
+        if os.path.exists(DIR_PATH + "/images"):
+            shutil.rmtree(DIR_PATH + "/images")
+
+    def upload_to_OS(self):
+        # Checking this, because OS is failing silently
+        env = os.environ.copy()
+        if not (
+            "OS_AUTH_URL" in env
+            and "OS_APPLICATION_CREDENTIAL_ID" in env
+            and "OS_APPLICATION_CREDENTIAL_SECRET" in env
+        ):
+            print("OS credentials missing in environment vars")
+            sys.exit(1)
```

This leaves several use cases out (e.g. clouds.yaml, service token) see all the env vars involved in this [docs page](https://manpages.ubuntu.com/manpages/focal/man1/openstack.1.html#environment%20variables).

I'd say either let this fail (annoying but ok if the script can be rerun only to reupload the image) so that OpenStack itself handles the issue (it has all the logic needed to handle the issue that we do not want to replicate here), or wrap this in a try-except block that shows the original message + your extra info (only if you want to add extra info).

---

I see it does not work like this. See [code](https://github.com/usegalaxy-eu/vgcn/pull/80/files#diff-b70bfb027a8ef3ec4ab07a53a3b8ba83e63334ba3ce5ca6d9d773900ed452e5eR409-R414). I don't know then, maybe let's skip this detail.

---

It is not nice, I know, but I was not successful with importing the `openstack` modules either (established a session and I was able to `openstack image list` but not to upload and since it did not threw any errors, I considered it to be too time consuming for now)